### PR TITLE
Transcription Metadata Tags

### PR DESCRIPTION
**Pull request by simensta**
Fixes Issue #2576. 

This PR introduces text tagging functionality for transcriptions, similar to the text tagging feature in [AnnoTate](https://anno.tate.org.uk/#/). Find the staging branch with sample workflow [here](https://transcription-metadata-tags.pfe-preview.zooniverse.org/projects/markb-panoptes/focus-testing-project/classify?workflow=2359).

A volunteer should be able to 1) highlight a portion of their transcription, 2) click the input associated with the desired tag, which surrounds the selection text with the metadata tag.

In the project builder, any project with access to the text tool has access to the three metadata tags. In a future iteration, the project builders will be able to create there own tags, instead of choosing between "Insertion", "Deletion", and "Unclear". The tags are stored on the task object as an array of strings under the key `text_task`.
text tag editor interface

### CODE REVIEW

**Comment by mcbouslog** [comment]
cool feature! i hit this problem awhile ago and wasn't sure what to do, hadn't tried AnnoTate yet, will be great to have in PFE as well.

few comments in-line, otherwise...

- i think since we've created TextTaskEditor we can delete [this line](https://github.com/zooniverse/Panoptes-Front-End/blob/transcription-metadata-tags/app/classifier/tasks/generic-editor.cjsx#L27) and one in the index i'll comment in-line on?
- toggle `required` property in editor (see Generic Editor)

File: app/classifier/tasks/text/index.cjsx [hunk 1]
```diff
@@ -1,6 +1,7 @@
 React = require 'react'
 {Markdown} = require 'markdownz'
 GenericTask = require '../generic'
+TextTaskEditor = require './editor'
 GenericTaskEditor = require '../generic-editor'
```
Comment (mcbouslog) [inline app/classifier/tasks/text/index.cjsx]: can be deleted?

File: app/classifier/tasks/text/index.cjsx [hunk 2]
```diff
@@ -63,9 +64,35 @@ module.exports = React.createClass
...
+    endTag = '[/' + textTag + ']'
+    textArea = this.refs.textInput
+    textAreaValue = textArea.value
+    selectionStart = textArea.selectionStart
+    selectionEnd = textArea.selectionEnd
+    textBefore = textAreaValue.substring(0, selectionStart)
+    if selectionStart is selectionEnd
+      textAfter = textAreaValue.substring(selectionStart, textAreaValue.length)
+      value = textBefore + startTag + endTag + textAfter
+      newAnnotation = Object.assign @props.annotation, {value}
```
Comment (mcbouslog) [inline app/classifier/tasks/text/index.cjsx]: can lines 90-91 and lines 97-98 be moved after the if/else conditional?

File: app/classifier/tasks/text/editor.cjsx [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  getDefaultProps: ->
+    workflow: {}
+    task: {}
+
+  tagExists:(tag) ->
+    if @props.task.text_tags
+      @props.task.text_tags.indexOf(tag) > -1
+
+  updateTags:(e)->
```
Comment (mcbouslog) [inline app/classifier/tasks/text/editor.cjsx]: not sure if properly captures possibilities or if really any improvement, maybe there's an even better option, but i think a different way to write could be:
```
value = e.target.value
checked = e.target.checked
text_tags = @props.task.text_tags ? []
if text_tags.indexOf(value) is -1 and checked
  text_tags.push(value)
else if !checked
  removalIndex = text_tags.indexOf(value)
  text_tags.splice(removalIndex,1);
@props.task.text_tags = text_tags
@props.onChange @props.task

File: app/classifier/tasks/text/editor.cjsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        @props.onChange @props.task
+      else if !checked
+        removalIndex = @props.task.text_tags.indexOf(value)
+        @props.task.text_tags.splice(removalIndex,1);
+        @props.onChange @props.task
+    else 
+      @props.task.text_tags = []
+      @props.task.text_tags.push(value) if checked
+      @props.onChange @props.task
+
```
Comment (mcbouslog) [inline app/classifier/tasks/text/editor.cjsx]: i think the `@props.onChange @props.task`'s can be moved to after if statement?

**Comment by simensta** [comment]
@mcbouslog, thanks for your feedback! Incorporating your input has definitely increased the readability of the code. Further thoughts on the new changes?

**Pull request merged**